### ISSUE TITLE
Pulseeffects for Debian 9 Stretch

### ISSUE THREAD

**Issue by palikacska** [wwmm/pulseeffects#683]
Hi there,

I've recently tried pulseeffects and I think it's really a niche for Linux systems, especially when it comes to "cheap plastic sounding" laptops and very silent audio tracks of any kind. However, I'm kind of disappointed that PulseEffects is deisgned to work only on the most recent distributions, regardless of the fact that the majority of audio plugins it uses is there for many years in Linux distributions to do their job. So I don't really understand why it would be technically impossible to compile pulseeffects against less recent libraries in order to work with older distributions (mostly LTS ones). I'm on Debian Stretch now and it's my main system. Unfortunately it's marked as unsupported (and too old) for pulseeffects.

I tried the Flatpak version. It seems to work and all effects are operational. However, the flatpak version is very slow in general, especially its start time and produces periodic hangs, freezes and slowdowns when adjusting knobs with the mouse wheel. It also has way too much latency (100-200 ms) which is noticeable at videos and online conferences and doesn't integrate to the system well. It also seems to be a bit bloated as it takes up at least 300 MB and ships an entire GNOME 3 which I don't need at all. I also tried pulseeffects on Debian Buster, adding mikhailnov's bionic repository to my sources.list. That works like a charm. However, Debian Buster is not yet an option for upgrade, due to a few major unresolved issues like [this one](https://bugs.chromium.org/p/chromium/issues/detail?id=1083669).

So what I'm trying to do now is building pulseeffects and linking all library dependencies (that are not found or too old on Debian Stretch) statically.

I would like to ask for help about what operating system (or docker environment) should I start with in the first place. I thought of Ubuntu Bionic or Debian Buster. I want pulseeffects to link to PulseAudio and GTK3 dynamically and statically to the rest. From dynamic libraries I would need the Stretch version during compilation. From static libraries the least compatible versions would do. Should I compile all of these libraries and its dependencies one by one, by hand? Or is there a shortcut like bulding only the Stretch versions of gtk3 and pulseaudio on Buster then link pulseeffects statically using pre-built buster and their *-dev packages? Is there a least painful way to do this?

**Comment by wwmm** [comment]
The plugins are only part of the problem. They alone do nothing. The first problem you might have is the gcc compiler version. I use modern c++ features when developing PulseEffects. The gcc 6.3 in Debian 9 might not have support to all of them.

The Pulseaudio 10 will definetely be a problem. It has bugs that directly impact PulseEffects. You need at least Pulseaudio 12 to have the best experience possible.

The Boost version in Debian Stretch will be a problem. You could try to revert a commit the I did to PulseEffects when Boost deprecated a few header usage. But I would not be surprised if other problems surface.

I do not remember now but there was a reason why GStreamer `1.12.5` was set as minimum. The version `1.10.4` in Stretch will probably cause a crash somewhere.

You can't dynamically link "to the rest". The plugins are dynamic as well as a few GStreamer libraries we have to use.

**Comment by wwmm** [comment]
I also do not remember now but I had a reason to set meson `0.40` as minimum. The version `0.37.1` will cause crash somewhere in the build process.

Honestly trying to replicate part of what flatpak does by yourself will be painful. Very painful.

**Comment by Digitalone1** [comment]
> However, I'm kind of disappointed that PulseEffects is deisgned to work only on the most recent distributions, regardless of the fact that the majority of audio plugins it uses is there for many years in Linux distributions to do their job.

That's not true if you consider the fact that git master or latest release only work with latest lsp-plugins released few month ago, otherwise you won't get compressor and equalizer.

Unfortunately, if the distribution don't ship them, it's not a PulseEffects fault.

**Comment by palikacska** [comment]
Sad to hear about that. :cry:

It means that I need a different approach here if I want to stay on Stretch.

What I would need the most is an equalizer and a maximizer (or limiter). Is there a way of dynamically loading them into pulseaudio without pulseeffects to be controllable via any other GUI (like pavucontrol)? The most difficult part is that I need dozens of pacmd/pactl commands to get LADSPA or other plugins to work with pulseaudio but there is no GUI or anything I could adjust parameters with (like threshold for maximizer/limiter and bands for EQ).

I'm sure you're more experienced about advanced audio features and applications in Linux, can you suggest me some alternatives if my stretch compilation project couldn't work?

I need a graphical equalizer to work with Pulseaudio and a graphical limiter / maximizer. The most important part that I want to turn them off from a UI with a few click instead of typing commands or writing scripts for pacmd. It should work dynamically like if I connect a HDMI monitor as audio output I would like to apply the EQ for that with one click. Any ideas on that?

**Comment by wwmm** [comment]
From a development point of view supporting old distributions is problematic and really frustrating. You see that a bug was fixed in the latest libraries. Like the one related to Pulseaudio. But you can't have the fix in these distributions. Sometimes like it was probably the GStreamer case a very useful feature that was not there before is implemented and using it will improve the code quality and make your life easier. But if want to support these distributions you can't have it.

I do not expect every distribution to be like the one I use(Arch Linux). But they should not take so long to update basic libraries. Specially when important bugs are fixed by upstream and when useful functionality is added.

**Comment by wwmm** [comment]
Unless playing everything through `vlc` is an option I think that the ladspa approach you have already tried and PulseEffects are the easiest ways to have an equalizer together with a compressor.

**Comment by wwmm** [comment]
`mpv` should also be able to apply equalization and compression. But just like `vlc` it  won't be a global solution you could apply to other apps.
